lib/puppet.rb

```diff
@@ -141,6 +143,12 @@ def self.run_mode
   # Load all of the settings.
   require 'puppet/defaults'
 
+  # Now that settings are loaded we have the code loaded to be able to issue
+  # deprecation warnings. Warn if we're on a deprecated ruby version.
+  if RUBY_VERSION < Puppet::OLDEST_RECOMMENDED_RUBY_VERSION
+    Puppet.deprecation_warning("Support for ruby version #{RUBY_VERSION} is deprecated and will be removed in a future release. See https://docs.puppet.com/puppet/latest/system_requirements.html#ruby for a list of supported ruby versions.")
+  end
```

Comment:
I was at a loss to figure out how to add a spec for this. It's at the top level within the `Puppet` module. I didn't see any other specs that tested similar behavior at the top level with module `Puppet` and settled for adding specs for the constant. Happy to add more though if others see a better way.